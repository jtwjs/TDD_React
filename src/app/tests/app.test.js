import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import renderer from "react-test-renderer";

import App from "../app";
import HabitPresenter from "../habit_presenter";

describe("App", () => {
  let presenter;
  beforeEach(() => {
    presenter = new HabitPresenter([
      { id: 1, name: "Reading", count: 1 },
      { id: 2, name: "Running", count: 0 },
      { id: 3, name: "Coding", count: 1 },
    ]);
  });

  it("renders", () => {
    const component = renderer.create(<App presenter={presenter} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  describe("Component", () => {
    beforeEach(() => {
      render(<App presenter={presenter} />);
    });

    it("counts only active habits", () => {
      const button = screen.getAllByRole("button", { name: "increase" })[0];
      userEvent.click(button);
      const count = screen.getAllByTestId("count")[0];
      expect(count.innerHTML).toBe("2");
    });

    it("adds new habit", () => {
      const newHabit = "whats";
      const input = screen.getByPlaceholderText("Habit");
      const button = screen.getByText("Add");
      userEvent.type(input, newHabit);
      userEvent.click(button);
      const addedName = screen.getAllByTestId("name")[3];
      expect(addedName.innerHTML).toBe(newHabit);
      const addedCount = screen.getAllByTestId("count")[3];
      expect(addedCount.innerHTML).toBe("0");
    });

    it("deletes an item", () => {
      const first = screen.getAllByRole("button", { name: "delete" })[0];
      userEvent.click(first);
      const next = screen.getAllByTestId("name")[0];
      expect(next.innerHTML).not.toBe("Reading");
    });

    it("increases the counter", () => {
      const button = screen.getAllByRole("button", { name: "increase" })[0];
      userEvent.click(button);
      const count = screen.getAllByTestId("count")[0];
      expect(count.innerHTML).toBe("2");
    });

    it("decreases the counter", () => {
      const button = screen.getAllByRole("button", { name: "decrease" })[2];
      userEvent.click(button);
      const count = screen.getAllByTestId("count")[2];
      expect(count.innerHTML).toBe("0");
    });

    it("resets all counters", () => {
      const button = screen.getByText("Reset All");
      userEvent.click(button);
      screen.getAllByTestId("count").forEach((count) => {
        expect(count.innerHTML).toBe("0");
      });
    });
  });
});
