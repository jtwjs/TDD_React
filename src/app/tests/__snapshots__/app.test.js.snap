// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders 1`] = `
Array [
  <nav
    className="navbar"
  >
    <i
      className="navbar-logo fas fa-leaf"
    />
    <span>
      Habit Tracker
    </span>
    <span
      className="navbar-count"
      data-testid="count"
    >
      2
    </span>
  </nav>,
  <div
    className="habits"
  >
    <form
      className="add-form"
      onSubmit={[Function]}
    >
      <input
        className="add-input"
        placeholder="Habit"
        type="text"
      />
      <button
        className="add-button"
      >
        Add
      </button>
    </form>
    <ul>
      <li
        className="habit"
      >
        <span
          className="habit-name"
          data-testid="name"
        >
          Reading
        </span>
        <span
          className="habit-count"
          data-testid="count"
        >
          1
        </span>
        <button
          aria-label="increase"
          className="habit-button habit-increase"
          onClick={[Function]}
        >
          <i
            className="fas fa-plus-square"
          />
        </button>
        <button
          aria-label="decrease"
          className="habit-button habit-decrease"
          onClick={[Function]}
        >
          <i
            className="fas fa-minus-square"
          />
        </button>
        <button
          aria-label="delete"
          className="habit-button habit-delete"
          onClick={[Function]}
        >
          <i
            className="fas fa-trash"
          />
        </button>
      </li>
      <li
        className="habit"
      >
        <span
          className="habit-name"
          data-testid="name"
        >
          Running
        </span>
        <span
          className="habit-count"
          data-testid="count"
        >
          0
        </span>
        <button
          aria-label="increase"
          className="habit-button habit-increase"
          onClick={[Function]}
        >
          <i
            className="fas fa-plus-square"
          />
        </button>
        <button
          aria-label="decrease"
          className="habit-button habit-decrease"
          onClick={[Function]}
        >
          <i
            className="fas fa-minus-square"
          />
        </button>
        <button
          aria-label="delete"
          className="habit-button habit-delete"
          onClick={[Function]}
        >
          <i
            className="fas fa-trash"
          />
        </button>
      </li>
      <li
        className="habit"
      >
        <span
          className="habit-name"
          data-testid="name"
        >
          Coding
        </span>
        <span
          className="habit-count"
          data-testid="count"
        >
          1
        </span>
        <button
          aria-label="increase"
          className="habit-button habit-increase"
          onClick={[Function]}
        >
          <i
            className="fas fa-plus-square"
          />
        </button>
        <button
          aria-label="decrease"
          className="habit-button habit-decrease"
          onClick={[Function]}
        >
          <i
            className="fas fa-minus-square"
          />
        </button>
        <button
          aria-label="delete"
          className="habit-button habit-delete"
          onClick={[Function]}
        >
          <i
            className="fas fa-trash"
          />
        </button>
      </li>
    </ul>
    <button
      className="habits-reset"
      onClick={[Function]}
    >
      Reset All
    </button>
  </div>,
]
`;
